Benchmarks can't run under .NET Core 3.1

Looks like the BenchmarkDotNet settings for the project can't run under .NET Core 3.1:

```
// Build Exception: Standard output:
   Determining projects to restore...
D:\Repositories\AppMetrics\src\Core\benchmarks\App.Metrics.Benchmarks.Runner\bin\Release\netcoreapp3.1\fcb23a49-0bb0-4e2f-9cb4-3779a5f35258\BenchmarkDotNet.Autogenerated.csproj : error NU1201: Project App.Metrics.Benchmarks is not compatible with netcoreapp2.1 (.NETCoreApp,Version=v2.1). Project App.Metrics.Benchmarks supports: netcoreapp3.1 (.NETCoreApp,Version=v3.1)
  Failed to restore D:\Repositories\AppMetrics\src\Core\benchmarks\App.Metrics.Benchmarks.Runner\bin\Release\netcoreapp3.1\fcb23a49-0bb0-4e2f-9cb4-3779a5f35258\BenchmarkDotNet.Autogenerated.csproj (in 312 ms).
  6 of 7 projects are up-to-date for restore.
```

The issue here appears to be an old BenchmarkDotNet dependency that doesn't properly support .NET Core 3.1

--- a/src/Core/benchmarks/App.Metrics.Benchmarks.Runner/App.Metrics.Benchmarks.Runner.csproj
+++ b/src/Core/benchmarks/App.Metrics.Benchmarks.Runner/App.Metrics.Benchmarks.Runner.csproj
@@ -23,8 +23,7 @@
   </ItemGroup>
 
   <ItemGroup>
-    <PackageReference Include="BenchmarkDotNet" Version="0.11.3" />
-    <PackageReference Include="BenchmarkDotNet.Core" Version="0.10.14" />
+    <PackageReference Include="BenchmarkDotNet" Version="0.12.1" />
   </ItemGroup>
 
 </Project>

--- a/src/Core/benchmarks/App.Metrics.Benchmarks/App.Metrics.Benchmarks.csproj
+++ b/src/Core/benchmarks/App.Metrics.Benchmarks/App.Metrics.Benchmarks.csproj
@@ -4,6 +4,7 @@
 
   <PropertyGroup>
     <IsPackable>false</IsPackable>
+    <LangVersion>7.3</LangVersion>
   </PropertyGroup>
 
   <PropertyGroup>
@@ -21,7 +22,7 @@
   <ItemGroup>
     <PackageReference Include="Microsoft.NET.Test.Sdk" />
     <PackageReference Include="xunit" />
-    <PackageReference Include="BenchmarkDotNet" Version="0.11.3" />
+    <PackageReference Include="BenchmarkDotNet" Version="0.12.1" />
     <PackageReference Include="xunit.runner.visualstudio">
       <PrivateAssets>all</PrivateAssets>
       <IncludeAssets>runtime; build; native; contentfiles; analyzers</IncludeAssets>

--- a/src/Core/benchmarks/App.Metrics.Benchmarks/Configs/DefaultConfig.cs
+++ b/src/Core/benchmarks/App.Metrics.Benchmarks/Configs/DefaultConfig.cs
@@ -4,6 +4,7 @@
 
 using BenchmarkDotNet.Configs;
 using BenchmarkDotNet.Diagnosers;
+using BenchmarkDotNet.Environments;
 using BenchmarkDotNet.Exporters;
 using BenchmarkDotNet.Jobs;
 
@@ -13,9 +14,9 @@ namespace App.Metrics.Benchmarks.Configs
     {
         public DefaultConfig()
         {
-            Add(Job.Core);
-            Add(MarkdownExporter.GitHub);
-            Add(new MemoryDiagnoser());
+            AddJob(Job.Default.WithRuntime(CoreRuntime.Core31));
+            AddExporter(MarkdownExporter.GitHub);
+            AddDiagnoser(MemoryDiagnoser.Default);
         }
     }
 }

--- a/src/Core/benchmarks/App.Metrics.Benchmarks/Support/BenchmarkTestExecutor.cs
+++ b/src/Core/benchmarks/App.Metrics.Benchmarks/Support/BenchmarkTestExecutor.cs
@@ -59,12 +59,12 @@ namespace App.Metrics.Benchmarks.Support
 
             if (!config.GetLoggers().OfType<OutputLogger>().Any())
             {
-                config = config.With(_output != null ? new OutputLogger(_output) : ConsoleLogger.Default);
+                config = config.AddLogger(_output != null ? new OutputLogger(_output) : ConsoleLogger.Default);
             }
 
             if (!config.GetColumnProviders().Any())
             {
-                config = config.With(DefaultColumnProviders.Instance);
+                config = config.AddColumnProvider(DefaultColumnProviders.Instance);
             }
 
             // Make sure we ALWAYS combine the Config (default or passed in) with any Config applied to the Type/Class
@@ -97,8 +97,8 @@ namespace App.Metrics.Benchmarks.Support
         private IConfig CreateSimpleConfig(OutputLogger logger = null)
         {
             return new SingleRunFastConfig()
-                .With(logger ?? (_output != null ? new OutputLogger(_output) : ConsoleLogger.Default))
-                .With(DefaultColumnProviders.Instance);
+                .AddLogger(logger ?? (_output != null ? new OutputLogger(_output) : ConsoleLogger.Default))
+                .AddColumnProvider(DefaultColumnProviders.Instance);
         }
     }
 }

--- a/src/Core/benchmarks/App.Metrics.Benchmarks/Support/BenchmarkTestRunner.cs
+++ b/src/Core/benchmarks/App.Metrics.Benchmarks/Support/BenchmarkTestRunner.cs
@@ -27,9 +27,9 @@ namespace App.Metrics.Benchmarks.Support
         {
             internal SingleRunFastConfig(ITestOutputHelper output)
             {
-                Add(Job.Dry);
-                Add(ConsoleLogger.Default);
-                Add(new OutputLogger(output));
+                AddJob(Job.Dry);
+                AddLogger(ConsoleLogger.Default);
+                AddLogger(new OutputLogger(output));
             }
         }
     }

--- a/src/Core/benchmarks/App.Metrics.Benchmarks/Support/TestConfigs.cs
+++ b/src/Core/benchmarks/App.Metrics.Benchmarks/Support/TestConfigs.cs
@@ -12,17 +12,17 @@ namespace App.Metrics.Benchmarks.Support
 #pragma warning disable SA1649, SA1402
     public class SingleRunFastConfig : ManualConfig
     {
-        public SingleRunFastConfig() { Add(Job.Dry); }
+        public SingleRunFastConfig() { AddJob(Job.Dry); }
     }
 
     public class SingleRunMediumConfig : ManualConfig
     {
-        public SingleRunMediumConfig() { Add(new Job(Job.Dry) { Run = { IterationCount = 5 } }); }
+        public SingleRunMediumConfig() { AddJob(new Job(Job.Dry) { Run = { IterationCount = 5 } }); }
     }
 
     public class ThroughputFastConfig : ManualConfig
     {
-        public ThroughputFastConfig() { Add(new Job(Job.Dry) { Run = { RunStrategy = RunStrategy.Throughput, IterationCount = 1 } }); }
+        public ThroughputFastConfig() { AddJob(new Job(Job.Dry) { Run = { RunStrategy = RunStrategy.Throughput, IterationCount = 1 } }); }
     }
 
     public class DiagnoserConfig : ManualConfig
@@ -30,7 +30,7 @@ namespace App.Metrics.Benchmarks.Support
         public DiagnoserConfig()
         {
             // Diagnosers need enough runs to collects the statistics!
-            Add(new Job { Run = { LaunchCount = 1, WarmupCount = 1, IterationCount = 50 } });
+            AddJob(new Job { Run = { LaunchCount = 1, WarmupCount = 1, IterationCount = 50 } });
         }
     }
 #pragma warning restore SA1649, SA1402

